mergin/test/test_client.py

```diff
@@ -483,3 +498,77 @@ def test_clone_project(mc):
     projects = mc.projects_list(flag='created')
     assert any(p for p in projects if p['name'] == cloned_project_name and p['namespace'] == API_USER)
+
+
+def test_available_storage_validation(mc, mc2):
+    """
+    Testing of storage limit - should not be applied for user pushing changes into project with different namespace.
+    This test also tests giving read and write access to another user. Additionally tests also uploading of big file.
+    """
+    test_project = 'test_available_storage_validation'
+    test_project_fullname = API_USER2 + '/' + test_project
+
+    # cleanups
+    project_dir = os.path.join(TMP_DIR, test_project, API_USER)
+    cleanup(mc, test_project_fullname, [project_dir])
+    cleanup(mc2, test_project_fullname, [project_dir])
+
+    try:
+        # create new (empty) project on server
+        mc2.create_project(test_project)
+
+        # Add writer access to another client
+        project_info = get_project_info(mc2, API_USER2, test_project)
+        access = project_info['access']
+        access['writersnames'].append(API_USER)
+        access['readersnames'].append(API_USER)
+        project_info['access'] = access
+        mc2.set_project_access(test_project_fullname, access)
+
+        # check writers access
+        project_info = get_project_info(mc2, API_USER2, test_project)
+        access = project_info['access']
+        assert API_USER in access['writersnames']
+
+        # download project
+        mc.download_project(test_project_fullname, project_dir)
+        mp = MerginProject(project_dir)
+
+        # get user_info about storage capacity
+        user_info = mc.user_info()
+        storage_remaining = user_info['storage'] - user_info['disk_usage']
+
+        # generate dummy data (remaining storage + extra 1024b)
+        dummy_data_path = project_dir + "/data"
+        file_size = storage_remaining + 1024
+        _generate_big_file(dummy_data_path, file_size)
+
+        # try to upload
+        mc.push_project(project_dir)
+
+        # Check project content
+        project_info = get_project_info(mc2, API_USER2, test_project)
+        assert project_info['meta']['files_count'] == 1
+        assert project_info['meta']['size'] == file_size
+
+    # clean project due to high storage requirement
+    finally:
+        cleanup(mc, test_project_fullname, [project_dir])
+        cleanup(mc2, test_project_fullname, [project_dir])
+
+
+def get_project_info(mc, namespace, project_name):
+    projects = mc.projects_list(flag='created')
+    test_project_list = [p for p in projects if p['name'] == project_name and p['namespace'] == namespace]
+    assert any(test_project_list)
+    return test_project_list[0]
+
+
+def _generate_big_file(filepath, size):
+    """
+    generate big binary file with the specified size in bytes
+    :param filepath: full filepath
+    :param size: the size in bytes
+    """
+    with open('%s'%filepath, 'wb') as fout:
+        fout.write(os.urandom(size))
```

Comment:
urandom is potentially slow because it generates high quality random numbers good for cryptography. We can simply use pseudo-random bytes, or simply a constant stream of e.g. 'A' characters. By the way, don't we have a utility function for that alerady, to be used to test large file uploads?